lib/opam_solve.ml

```diff
@@ -121,99 +158,265 @@ let fixed_packages ~local_packages ~pin_depends =
+
+  let build_context ~build_only ~allow_jbuilder ~ocaml_version ~local_packages
+      ~pin_depends ~target_packages input =
+    let open Result.O in
+    let install_test_deps_for =
+      if build_only then OpamPackage.Name.Set.empty else target_packages
+    in
+    let constraints = constraints ~ocaml_version in
+    let+ fixed_packages = fixed_packages ~local_packages ~pin_depends in
+    Context.create ~install_test_deps_for ~allow_jbuilder ~constraints
+      ~fixed_packages input
+
+  let calculate_raw ~local_packages ~target_packages context =
+    let target_packages_names = OpamPackage.Name.Set.elements target_packages in
+    let allow_compiler_variants = depend_on_compiler_variants local_packages in
+    let request = request ~allow_compiler_variants target_packages_names in
+    let result = Solver.solve context request in
+    match result with
+    | Error e -> Error (`Diagnostics e)
+    | Ok selections ->
+        let packages = Solver.packages_of_result selections in
+        let deps =
+          List.filter
+            ~f:(fun pkg ->
+              let name = OpamPackage.name pkg in
+              let in_local_packages =
+                OpamPackage.Name.Map.mem name local_packages
+              in
+              not in_local_packages)
+            packages
+        in
+        Ok deps
+
+  type diagnostics = Solver.diagnostics
+
+  let diagnostics_message ~verbose diagnostics =
+    `Msg (Solver.diagnostics ~verbose diagnostics)
+
+  module Pkg_map = Solver.Solver.Output.RoleMap
+
+  let no_version_builds_with_dune component =
+    match Solver.Diagnostics.Component.selected_impl component with
+    | Some _ -> false
+    | None -> (
+        let rejects, _reason = Solver.Diagnostics.Component.rejects component in
+        match rejects with
+        | [] -> false
+        | _ ->
+            List.for_all
+              ~f:(fun (_, reason) ->
+                match reason with
+                | `Model_rejection Context.Non_dune -> true
+                | _ -> false)
+              rejects)
+
+  let not_buildable_with_dune diagnostics =
+    let rolemap = Solver.diagnostics_rolemap diagnostics in
+    Pkg_map.fold
+      (fun pkg component acc ->
+        match no_version_builds_with_dune component with
+        | false -> acc
+        | true -> pkg :: acc)
+      rolemap []
+    |> List.filter_map ~f:Solver.package_name
+
+  let get_opam_info ~context pkg =
+    match Context.opam_file context pkg with
+    | Ok opam_file -> Opam.Package_summary.from_opam ~pkg opam_file
+    | Error (`Msg msg) ->
+        (* If we're calling this function on a package, it means it has been
+           returned as part of the solver solution and therefore should correspond
+           to a valid candidate. *)
+        Logs.debug (fun l ->
+            l "Could not retrieve opam file for %a: %s" Opam.Pp.package pkg msg);
+        assert false
+
+  (* TODO catch exceptions and turn to error *)
```

Comment:
TODO :)

Reply:
This was already present before this PR, this is just moved code.